python/lsst/obs/base/gen3/repoConverter.py

```diff
@@ -0,0 +1,342 @@
+        if filePathParser is None:
+            filePathParser = FilePathParser.fromMapping(mapper.mappings[datasetTypeName])
+        self.filePathParser = filePathParser
+        if translator is None:
+            translator = Translator.makeMatching(filePathParser.datasetType, baseDataId, skyMap=skyMap)
+        self.translator = translator
+        self.datasetType = DatasetType(datasetTypeName, dimensions=self.translator.dimensions,
+                                       storageClass=storageClass)
+        self.datasetType.normalize(universe=universe)
+
+    def apply(self, fileNameInRoot):
+        """Extract a Gen3 data ID from the given filename,
+
+        Parameters
+        ----------
+        fileNameInRoot : `str`
+            Filename relative to a Gen2 data repository root.
+
+        Returns
+        -------
+        dataId : `lsst.daf.butler.DataId` or `None`
+            The Gen3 data ID, or `None` if the file was not recognized as an
+            instance of the extractor's dataset type.
+        """
+        gen2id = self.filePathParser(fileNameInRoot)
+        if gen2id is None:
+            return None
+        return DataId(self.translator(gen2id), dimensions=self.datasetType.dimensions)
+
+
+class RepoConverter:
+    """A helper class that ingests (some of) the contents of a Gen2 data
+    repository into a Gen3 data repository.
+
+    Parameters
+    ----------
+    root : `str`
+        Root of the Gen2 data repository.
+    universe : `lsst.daf.butler.DimensionUniverse`
+        Object containing all dimension definitions.
+    baseDataId : `dict`
+        Key-value pairs that may need to appear in the Gen3 data ID, but can
+        never be inferred from a Gen2 filename.  This should always include
+        the instrument name (even Gen3 data IDs that don't involve the
+        instrument dimension have instrument-dependent Gen2 filenames) and
+        should also include the skymap name in order to process any data IDs
+        that involve tracts or patches.
+    mapper : `lsst.obs.base.CameraMapper`, optional
+        Object that defines Gen2 filename templates.  Will be identified,
+        imported, and constructed from ``root`` if not provided.
+    skyMap : `lsst.skymap.BaseSkyMap`, optional
+        SkyMap that defines tracts and patches.  Must be provided in order to
+        provess datasets with a ``patch`` key in their data IDs.
+    """
+
+    COADD_NAMES = ("deep", "goodSeeing", "dcr")
+    REPO_ROOT_FILES = ("registry.sqlite3", "_mapper", "repositoryCfg.yaml",
+                       "calibRegistry.sqlite3", "_parent")
+
+    def __init__(self, root, *, universe, baseDataId, mapper=None, skyMap=None):
+        self.root = root
+        if mapper is None:
+            # Shush spurious log messages from Gen2 Mapper classes.
+            with loggedAt("CameraMapper", Log.ERROR):
+                with loggedAt("HscMapper", Log.ERROR):
```

Comment:
Do we really need to do this to quiet those log messages? Maybe we can just lower their log levels of those messages? Or if this feature is often needed maybe it can be added to lsst.log?

Reply:
At least the `HscMapper` messages are worthwhile warnings in other contexts, and I think some of the `CameraMapper` ones are too.  The difference is that I'm not being particular about how I initialize the mapper here because I don't care about parent lookup, because all need to get from the mapper are the templates for each dataset type.

But I think moving this to lsst.log is a good idea; I'll do that.